File: BasicML/basicml/optim/rmsprop.py
from .optimizer             import Optimizer
from typing                 import Optional
from basicml.regularization import Regularizer
from basicml.tensor         import Tensor
import numpy as np

class RMSProp(Optimizer):
    def __init__(self,
                 parameters : list[Tensor],
                 lr         : float,
                 regularizer: Optional[Regularizer] = None,
                 rho        : float                 = 0.9,
                 eps        : float                 = 1e-8):
        super().__init__(parameters, lr, regularizer)
        self.rho        = rho
        self.eps        = eps
        self.mean_sq    = [np.zeros_like(p.data) for p in self.parameters]

    def step(self):
        for i, param in enumerate(self.parameters):
            if param.requires_grad and param.grad is not None:
                grad = param.grad
                if self.regularizer is not None:
                    grad = grad + self.regularizer.grad(param)
                self.mean_sq[i] = self.rho * self.mean_sq[i] + (1.0 - self.rho) * grad ** 2
                param.data -= self.lr * grad / (np.sqrt(self.mean_sq[i]) + self.eps)

    def zero_grad(self):
        for param in self.parameters:
            param.zero_grad()

File: AdvanceML/src/optim/rmsprop.cpp
#include "advanceml/optim/rmsprop.hpp"

#include <cmath>

namespace advanceml {

RMSprop::RMSprop(std::vector<Tensor> parameters, float learning_rate, float rho, float eps)
    : parameters_(std::move(parameters)), learning_rate_(learning_rate), rho_(rho), eps_(eps) {
    mean_sq_.reserve(parameters_.size());
    for (const Tensor& param : parameters_) {
        mean_sq_.emplace_back(param.data().size(), 0.0f);
    }
}

void RMSprop::step() {
    for (size_t i = 0; i < parameters_.size(); ++i) {
        Tensor& param = parameters_[i];
        if (!param.has_grad()) {
            continue;
        }
        Tensor grad = param.grad();
        std::vector<float>& mean_sq = mean_sq_[i];
        for (size_t j = 0; j < param.data().size(); ++j) {
            const float g = grad.data()[j];
            mean_sq[j] = rho_ * mean_sq[j] + (1.0f - rho_) * g * g;
            param.data()[j] -= learning_rate_ * g / (std::sqrt(mean_sq[j]) + eps_);
        }
    }
}

void RMSprop::zero_grad() {
    for (Tensor& param : parameters_) {
        param.zero_grad();
    }
}

}  // namespace advanceml
